feat(langchain): middleware support in `create_agent`

sydney-runkle opened the pull request:
## Overview

Adding new `AgentMiddleware` primitive that supports `before_model`, `after_model`, and `prepare_model_request` hooks.

This is very exciting! It makes our `create_agent` prebuilt much more extensible + capable. Still in alpha and subject to change.

This is different than the initial [implementation](https://github.com/langchain-ai/langgraph/tree/nc/25aug/agent) in that it:
* Fills in gaps w/ missing features, for ex -- new structured output, optionality of tools + system prompt, sync and async model requests, provider builtin tools
* Exposes private state extensions for middleware, enabling things like model call tracking, etc
* Middleware can register tools
* Uses a `TypedDict` for `AgentState` -- dataclass subclassing is tricky w/ required values + required decorators
* Addition of `model_settings` to `ModelRequest` so that we can pass through things to bind (like cache kwargs for anthropic middleware)

## TODOs

### top prio
- [x] add middleware support to existing agent
- [x] top prio middlewares
  - [x] summarization node
  - [x] HITL
  - [x] prompt caching
 
other ones
- [x] model call limits
- [x] tool calling limits
- [ ] usage (requires output state)

### secondary prio
- [x] improve typing for state updates from middleware (not working right now w/ simple `AgentUpdate` and `AgentJump`, at least in Python)
- [ ] add support for public state (input / output modifications via pregel channel mods) -- to be tackled in another PR
- [x] testing!

### docs
See https://github.com/langchain-ai/docs/pull/390
- [x] high level docs about middleware
- [x] summarization node
- [x] HITL
- [x] prompt caching

## open questions

Lots of open questions right now, many of them inlined as comments for the short term, will catalog some more significant ones here.

sydney-runkle committed:
first pass at new agent w/ middleware

sydney-runkle committed:
initial middlewares + testing

sydney-runkle committed:
some linting

sydney-runkle committed:
adding support to existing agent

sydney-runkle committed:
model settings and prompt caching

sydney-runkle committed:
enhancing prompt caching

sydney-runkle committed:
refactored summarization node

hwchase17 commented on libs/langchain_v1/langchain/agents/middleware/summarization.py:
```diff
@@ -0,0 +1,272 @@
+
+    def __init__(
+        self,
+        model: BaseChatModel,
```
could we also accept string here? i think in general, we should try to accept string and then use init_chat_model a bunch

hwchase17 commented on libs/langchain_v1/langchain/agents/middleware/summarization.py:
```diff
@@ -0,0 +1,272 @@
+        messages = state["messages"]
+        self._ensure_message_ids(messages)
+
+        total_tokens = self.token_counter(messages)
```
do we want to cache this or something? recomputing kind of annoying?

hwchase17 commented on libs/langchain_v1/langchain/agents/middleware/summarization.py:
```diff
@@ -0,0 +1,272 @@
+        ):
+            return None
+
+        system_message, conversation_messages = self._split_system_message(messages)
```
i think we probably want to assume that system message is not in messages right?

hwchase17 commented on libs/langchain_v1/langchain/agents/middleware/summarization.py:
```diff
@@ -0,0 +1,272 @@
+        )
+
+        summary = self._create_summary(messages_to_summarize)
+        updated_system_message = self._build_updated_system_message(system_message, summary)
```
mmmm no, should probably not assume message is here?

hwchase17 commented on libs/langchain_v1/langchain/agents/middleware/human_in_the_loop.py:
```diff
@@ -0,0 +1,101 @@
+                }
+                approved_tool_calls.append(new_tool_call)
+            elif response["type"] == "ignore":
+                # NOTE: does not work with multiple interrupts
```
if it doesnt work with multiple interrupts... can we just raise error rather than silently failing?

hwchase17 committed:
cr

hwchase17 committed:
cr

sydney-runkle committed:
chore(langchain): fixing typing + cleanup namespace for middleware (#32837)

Main change here is the removal of `AgentJump`.

Typing things as `AgentJump` doesn't really make sense, given that we
often want to perform an agent jump **and** update the state of the
graph, potentially including updates to state extensions from a given
middleware.

For example, previously we had:

```py
def after_model(state: AgentState) -> AgentUpdate | AgentJump | None:
    ...
```

But what if my middleware adds a `model_calls` attribute? `AgentJump`
has no way of inheriting that newly defined state, but I still might
want to do something like:

```py
def after_model(state: AgentState) -> AgentUpdate | AgentJump | None:
    return {'model_calls': X, 'jump_to': '__end__'}
```

Then the return type `AgentUpdate | AgentJump | None` doesn't represent
the return value.

There's no great way to type partial state updates yet, hence we use
`dict[str, Any]`. This is a Python limitation.

I've removed the conditional logic for signature parsing on `AgentJump`
and we just always add conditional edges for middleware nodes now.

Other things:
* adding docstrings + other linting fixes (sort of a general cleanup of
this module).
* removing tool calling / model calling limits as those applications
aren't fully fleshed out / we want to just publish these 3 middlewares
first.

sydney-runkle committed:
minor fixes

sydney-runkle committed:
moving around types

sydney-runkle committed:
the beginning of testing

sydney-runkle committed:
removing scratch files

sydney-runkle committed:
chore(langchain): prepare model request via func call not node (#32838)

sydney-runkle committed:
tests w/ claude as a starting point

sydney-runkle committed:
fix imports

sydney-runkle committed:
version bump

sydney-runkle committed:
import top level

sydney-runkle committed:
linting

sydney-runkle committed:
Merge branch 'master' into sr/new-agent
